crates/bevy_reflect/src/utility.rs

```diff
@@ -0,0 +1,149 @@
+///       TypeInfo::Value(info)
+///     })
+///   }
+/// }
+/// #
+/// # unsafe impl<T: Reflect> Reflect for Foo<T> {
+/// #   fn type_name(&self) -> &str { todo!() }
+/// #   fn get_type_info(&self) -> &'static TypeInfo { todo!() }
+/// #   fn any(&self) -> &dyn Any { todo!() }
+/// #   fn any_mut(&mut self) -> &mut dyn Any { todo!() }
+/// #   fn as_reflect(&self) -> &dyn Reflect { todo!() }
+/// #   fn as_reflect_mut(&mut self) -> &mut dyn Reflect { todo!() }
+/// #   fn apply(&mut self, value: &dyn Reflect) { todo!() }
+/// #   fn set(&mut self, value: Box<dyn Reflect>) -> Result<(), Box<dyn Reflect>> { todo!() }
+/// #   fn reflect_ref(&self) -> ReflectRef { todo!() }
+/// #   fn reflect_mut(&mut self) -> ReflectMut { todo!() }
+/// #   fn clone_value(&self) -> Box<dyn Reflect> { todo!() }
+/// # }
+/// ```
+///
+/// [`once_cell`]: https://docs.rs/once_cell/latest/once_cell/
+/// [`non_generic`]: TypeInfoCell::non_generic
+/// [`generic`]: TypeInfoCell::generic
+pub struct TypeInfoCell(TypeInfoCellType);
+
+impl TypeInfoCell {
+    /// Initialize a [`TypeInfoCell`] for non-generic types.
+    pub const fn non_generic() -> Self {
+        Self(TypeInfoCellType::NonGeneric(OnceBox::new()))
+    }
+
+    /// Initialize a [`TypeInfoCell`] for generic types.
+    pub const fn generic() -> Self {
+        Self(TypeInfoCellType::Generic(OnceBox::new()))
+    }
+
+    /// Returns a reference to the [`TypeInfo`] stored in the cell.
+    ///
+    /// If there is no [`TypeInfo`] found, a new one will be generated from the given function.
+    ///
+    /// # Generics
+    ///
+    /// Generic types, such as `Vec<T>`, store a mapping of [`TypeIds`] to [`TypeInfos`]. This
+    /// method will then return the correct [`TypeInfo`] reference for the given type `T`.
+    ///
+    /// [`TypeIds`]: std::any::TypeId
+    /// [`TypeInfos`]: TypeInfo
+    pub fn get_or_insert<T, F>(&self, f: F) -> &TypeInfo
+    where
+        T: Any + ?Sized,
+        F: FnOnce() -> TypeInfo,
+    {
+        match &self.0 {
+            TypeInfoCellType::NonGeneric(once) => once.get_or_init(|| Box::new(f())),
+            TypeInfoCellType::Generic(once) => {
+                let type_id = TypeId::of::<T>();
+                let mapping = once.get_or_init(|| Box::new(RwLock::default()));
+                if let Some(info) = mapping.read().get(&type_id) {
+                    return info;
+                }
+
+                mapping.write().entry(type_id).or_insert_with(|| {
+                    // We leak here in order to obtain a `&'static` reference.
+                    // Otherwise, we won't be able to return a reference due to the `RwLock`.
+                    // This should be okay, though, since we expect it to remain statically
+                    // available over the course of the application.
+                    Box::leak(Box::new(f()))
+                })
+            }
+        }
+    }
+}
+
+enum TypeInfoCellType {
+    NonGeneric(OnceBox<TypeInfo>),
```

Comment:
I went with the enum so that users only need to know of `TypeInfoCell`'s existence rather than a `GenericTypeInfoCell` and `NonGenericTypeInfoCell`. But either is fine.

I'll split this into two `***TypeInfoCell` structs.